Mustachio/Tokenizer.cs

```diff
@@ -190,35 +206,54 @@ public static IEnumerable<TokenPair> Tokenize(string templateString)
-                    parseErrors.Add(new IndexedParseException(unclosedScope.location,
+                    parseErrors.Add(new IndexedParseException(sourceName, unclosedScope.location,
                         "A scope block to the following path was opened but not closed: '{0}', please close it using the appropriate syntax.",
                         unclosedScope.scope));
                 }
             }
             #endregion
 
-            //We want to throw an aggregate template exception, but in due time.
-            if (parseErrors.Any())
+            return new TokenizeResult { Tokens = tokens, Errors = parseErrors };
+        }
+
+        private static bool ExpandTokens(IEnumerable<TokenExpander> expanders, Match m, ParsingOptions options,
+            ref List<TokenTuple> tokens, ref List<IndexedParseException> parseErrors)
+        {
+            var expander = expanders.FirstOrDefault(e => e.RegEx.IsMatch(m.Value));
+
+            if (expander == null)
+            {
+                return false;
+            }
+
+            if (expander.Renderer != null)
+            {
+                tokens.Add(new TokenTuple(TokenType.Custom, m.Value, expander.Renderer));
+            }
+
+            if (expander.ExpandTokens != null)
             {
-                var innerExceptions = parseErrors.OrderBy(k => k.LineNumber).ThenBy(k => k.CharacterOnLine).ToArray();
-                throw new AggregateException(innerExceptions);
+                var tokenizeResult = expander.ExpandTokens(m.Value, options);
+                tokens.AddRange(tokenizeResult.Tokens);
+                parseErrors.AddRange(tokenizeResult.Errors);
             }
-            yield break;
+            
+            return true;
         }
 
         /// <summary>
-        /// Specifies combnations of paths that don't work.
+        /// Specifies combinations of paths that don't work.
         /// </summary>
         private static readonly Regex _negativePathSpec = new Regex("([.]{3,})|([^\\w./_]+)|((?<![.]{2})[/])|([.]{2,}($|[^/]))", RegexOptions.Singleline | RegexOptions.Compiled);
 
-        private static string Validated(string token, string content, int index, ref int[] lines, ref List<IndexedParseException> exceptions)
+        private static string Validated(string token, string content, int index, string sourceName, ref int[] lines, ref List<IndexedParseException> exceptions)
         {
             token = token.Trim();
 
             if (_negativePathSpec.Match(token).Success)
             {
                 var location = HumanizeCharacterLocation(content, index, ref lines);
-                exceptions.Add(new IndexedParseException(location, "The path '{0}' is not valid. Please see documentation for examples of valid paths.", token));
+                exceptions.Add(new IndexedParseException(sourceName, location, "The path '{0}' is not valid. Please see documentation for examples of valid paths.", token));
```

Comment:
We don't do any string formatting ourselves for those exceptions within Mustachio. The SourceName property is exposed in the thrown exceptions. I pondered whether that should be set to "unknown" instead of null, but I slightly prefer returning null and letting the "clients" check against that instead of an "unknown" string. I don't have a strong opinion though.

Reply:
We should just return it as `null`, which in the case of loading it from a DB field might be better than "unknown" -- I was simply suggesting this if we formatted a message, but otherwise, no need.